src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,142 @@
+            grow();
+        }
+        Node<K, V> currentNode = table[bucketIndex];
+        while (currentNode != null) {
+            if (Objects.equals(currentNode.key, key)) {
+                putValueInExistingNode(currentNode, key, value);
+                return;
+            }
+            if (currentNode.next == null) {
+                putFirstNodeInBucket(currentNode, key, value);
+                return;
+            }
+            currentNode = currentNode.next;
+        }
+        putNodeInBucket(key, value, bucketIndex);
+    }
+
+    private void putNodeInBucket(K key, V value, int bucketIndex) {
+        table[bucketIndex] = new Node<>(key, value, null);
+        size++;
+    }
 
+    private void putFirstNodeInBucket(Node<K, V> node, K key, V value) {
+        node.next = new Node<>(key, value, null);
+        size++;
+    }
+
+    private void putValueInExistingNode(Node<K, V> node, K key, V value) {
+        node.value = value;
+    }
+
+    private void grow() {
+        size = 0;
+        Node<K, V>[] bufferTable = table;
+        table = new Node[table.length * GROW_COEFFICIENT];
+        for (Node<K, V> node : bufferTable) {
+            while (node != null) {
+                put(node.key, node.value);
+                node = node.next;
+            }
+        }
+        threshold *= GROW_COEFFICIENT;
     }
 
     @Override
     public V getValue(K key) {
+        int bucketIndex = getBucketByKey(key);
+        Node<K, V> currentNode = table[bucketIndex];
+        while (currentNode != null) {
+            if (Objects.equals(key, currentNode.key)) {
+                return currentNode.value;
+            }
+            currentNode = currentNode.next;
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    @Override
+    public boolean containsKey(K key) {
+        int bucketIndex = getBucketByKey(key);
+        Node<K, V> currentNode = table[bucketIndex];
+        while (currentNode != null) {
+            if (Objects.equals(key, currentNode.key)) {
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean containsValue(V value) {
+        for (Node<K, V> node : table) {
+            Node<K, V> currentNode = node;
+            while (currentNode != null) {
+                if (Objects.equals(value, currentNode.value)) {
+                    return true;
+                }
+                currentNode = currentNode.next;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public void remove(K key) {
+        int bucketIndex = getBucketByKey(key);
+        Node<K, V> currentNode = table[bucketIndex];
+        while (currentNode != null) {
+            if (Objects.equals(key, currentNode.key)) {
+                table[bucketIndex] = null;
+                size--;
+                return;
```

Comment:
what happens if we delete an element in the middle of the list?

Reply:
fixes this method.